### PR TITLE
[charts] Introduce dataset to simplify plot of data from API

### LINKED ISSUE #9283: [charts] Allows bar/line charts to process an array of objects
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Summary 💡

For now, the data are provided as follows:
- `series` get a `data` property providing the y values.
- `xAxis` get a `data` property providing the x values.

But most of the time you receive data from API as follows

```js
[
    {"date": "2023-06-01", "value": 100},
    {"date": "2023-06-02", "value": 105},
    {"date": "2023-06-03", "value": 98},
    ...
]
```

The idea would be to for the following:

```jsx
<BarCharts
  data={fetchedData}
  series={[ { dataKey: "value", ... }	]}
  xAxis={[ { dataKey: "date", scaleType: 'time' } ]}
/>
```

### Examples 🌈

_No response_

### Motivation 🔦

Smoother server-side integration for simple/common cases. Currently, users most likely have to pre-process any data coming from API endpoints to have them separate in two different arrays (for X and Y axis).

### Order ID 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by alexfauquette**
Fix #9283

I did a first draft of the bar chart to demonstrate how it would work

Docs and example available here: https://deploy-preview-9774--material-ui-x.netlify.app/x/react-charts/bars/#using-a-dataset

TODO:

- [x] axes
- [x] bar chart
- [x] line chart
- [ ] scatter chart (unsure)

For the pie chart, it does not seem to be relevant since its data prop is already an array of objects.

### CODE REVIEW

File: packages/x-charts/src/models/axis.ts [hunk 1]
```diff
@@ -142,6 +142,10 @@ export type AxisConfig<S extends ScaleName = ScaleName, V = any> = {
   min?: number;
   max?: number;
   data?: V[];
+  /**
+   * The key used to retrive data from the dataset prop.
```
Comment (LukasTy) [inline packages/x-charts/src/models/axis.ts:146]:
```suggestion
   * The key used to retrieve data from the dataset prop.
```

File: docs/data/charts/bars/bars.md [hunk 2]
```diff
@@ -15,6 +15,16 @@ This axis might have `scaleType='band'` and its `data` should have the same leng
 
 {{"demo": "BasicBars.js"}}
 
+### Using a dataset
+
+If your data are stored in an array of objects, you can use the `dataset` helper prop.
```
Comment (LukasTy) [inline docs/data/charts/bars/bars.md:20]:
```suggestion
If your data is stored in an array of objects, you can use the `dataset` helper prop.
```

File: docs/data/charts/bars/bars.md [hunk 3]
```diff
@@ -15,6 +15,16 @@ This axis might have `scaleType='band'` and its `data` should have the same leng
...
+It accepts an array of objects such as `dataset={[{x: 1, y: 32}, {x: 2, y: 41}, ...]}`.
+
+You can reuse those data to define the series and axis thanks to `dataKey` properties.
```
Comment (LukasTy) [inline docs/data/charts/bars/bars.md:23]:
```suggestion
You can reuse this data when defining the series and axis thanks to the `dataKey` property.
```

File: packages/x-charts/src/BarChart/formatter.ts [hunk 4]
```diff
@@ -2,35 +2,55 @@ import { stack as d3Stack } from 'd3-shape';
...
+          d3Dataset.push({ [id]: value });
+        } else {
+          d3Dataset[index][id] = value;
+        }
+      });
+    } else if (dataset === undefined) {
+      throw new Error(
+        [
+          `MUI: bar series with id='${id}' has no data.`,
+          'Either provide a data property to the series or use dataset props.',
```
Comment (LukasTy) [inline packages/x-charts/src/BarChart/formatter.ts:27]:
```suggestion
          'Either provide a data property to the series or use the dataset prop.',
```

**Review by LukasTy — COMMENTED** [review]
Looks like a very handy improvement! 👍

File: docs/data/charts/bars/bars.md [hunk 5]
```diff
@@ -15,6 +15,16 @@ This axis might have `scaleType='band'` and its `data` should have the same leng
 
 {{"demo": "BasicBars.js"}}
 
+### Using a dataset
+
+If your data is stored in an array of objects, you can use the `dataset` helper prop.
+It accepts an array of objects such as `dataset={[{x: 1, y: 32}, {x: 2, y: 41}, ...]}`.
+
+You can reuse this data when defining the series and axis thanks to the `dataKey` property.
```
Comment (richbustos) [inline docs/data/charts/bars/bars.md:23]:
```suggestion
You can reuse this data when defining the series and axis, thanks to the `dataKey` property.
```

File: docs/data/charts/lines/lines.md [hunk 6]
```diff
@@ -18,6 +18,18 @@ To modify the x values, you should provide a `xAxis` with data property.
 
 {{"demo": "BasicLineChart.js"}}
 
+### Using a dataset
+
+If your data is stored in an array of objects, you can use the `dataset` helper prop.
+It accepts an array of objects such as `dataset={[{x: 1, y: 32}, {x: 2, y: 41}, ...]}`.
+
+You can reuse this data when defining the series and axis thanks to the `dataKey` property.
```
Comment (richbustos) [inline docs/data/charts/lines/lines.md:26]:
```suggestion
You can reuse this data when defining the series and axis, thanks to the `dataKey` property.
```

**Comment by richbustos** [comment]
Screen Shot 2023-07-26 at 3 15 08 PM

Nothing major, but what do you think of adding a line break before "For example...". 

I think it would be visually please if it was on a new line

**Review by LukasTy — APPROVED** [review]

**Pull request merged**